feat: make sshnpd execute ssh command via Process.start

gkc opened the pull request:
**- What I did**
- feat: For perf reasons, by default we will have SSHNPDImpl directly execute the ssh command on the host rather than using SSHClient
  - writes privatekey which it received earlier from sshnp into a tmp file which is later deleted
  - added a bunch of comments explaining how the command is constructed and why
- refactor: remove _sshPublicKey instance variable from SSHNPDImpl and make it a local variable in the `case 'sshpublickey'` in the notification handler
- feat: Direct AtSignLogger log messages to stdErr for sshnpd, sshnp and sshrvd via `AtSignLogger.defaultLoggingHandler = AtSignLogger.stdErrLoggingHandler;`
- feat: Added command-line option `--ssh-client` which can currently have values '/usr/bin/ssh' (the default) or 'pure-dart'. If set to `pure-dart` then we'll do it using SSHClient as we used to.
- feat: added more command-line options to the ssh command which sshnp generates and prints to stdout
  - `-o StrictHostKeyChecking=accept-new` to silently accept new, but still check existing
  - `-o IdentitiesOnly=yes` to use only the identity in the specified identity file

gkc committed:
feat: Direct AtSignLogger log messages to stdErr for sshnpd, sshnp and sshrvd via `AtSignLogger.defaultLoggingHandler = AtSignLogger.stdErrLoggingHandler;`

gkc committed:
feat: For perf reasons, have SSHNPDImpl directly execute the ssh command on the host rather than using SSHClient
- writes privatekey which it received earlier from sshnp into a tmp file which is later deleted
- added a bunch of comments explaining how the command is constructed
refactor: remove _sshPublicKey instance variable from SSHNPDImpl and make it a local variable in the `case 'sshpublickey'` in the notification handler

gkc commented:
@cconstab My local manual tests worked but the e2e tests are failing with 'Host key verification failed', because of something I'm getting wrong. Please take a look at how I'm constructing the ssh command - a fresh pair of eyes might see what the problem is straight away.

gkc commented:
I believe the tests are failing because SSHClient does not do host key verification, but ssh command by default does 

https://github.com/TerminalStudio/dartssh2/issues/51

Fix would be to add this to the ssh command args

-o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null

gkc commented:
Fix would be this https://askubuntu.com/a/87452

gkc committed:
fix: Add `-o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null` to the ssh command

gkc committed:
build: Removed dependency on dartssh2 and ssh_key packages

gkc committed:
build: Re-added dependency on dartssh2 and ssh_key packages to test a hypothesis.

gkc committed:
Merge remote-tracking branch 'origin/trunk' into gkc-exec-ssh-on-host

gkc committed:
build: Re-added dependency on dartssh2 and ssh_key packages to test a hypothesis.

gkc committed:
build: Re-removed dependency on dartssh2 and ssh_key packages

gkc commented:
Applied the fix mentioned in the previous comment, and now only one of the e2e tests is failing, but it's failing in a different way. 
```
130 sshnpd  | SHOUT|2023-07-23 18:19:51.546987| sshnpd |197de568-697a-4851-a789-9e82a722e5cb | Non-zero exit code from /usr/bin/ssh atsign@10.1.0.38 -p 40749 -t -t -i /tmp/.pem-436bdbbb-5052-4cb8-bd16-80ab0f86c376 -R 34925:localhost:22 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null sleep 15 
131 sshnpd  | SHOUT|2023-07-23 18:19:51.548265| sshnpd |197de568-697a-4851-a789-9e82a722e5cb | stdout   :  
132 sshnpd  | SHOUT|2023-07-23 18:19:51.548293| sshnpd |197de568-697a-4851-a789-9e82a722e5cb | stderr   : Warning: Permanently added '[10.1.0.38]:40749' (ED25519) to the list of known hosts.
133 sshnpd  | Load key "/tmp/.pem-436bdbbb-5052-4cb8-bd16-80ab0f86c376": error in libcrypto
134 sshnpd  | atsign@10.1.0.38: Permission denied (publickey).
```

I *think* the repeat failure of the local-to-local test is an artefact of the test rig

All of the other tests which are running sshnpd from the branch (local) are passing  (i.e. trunk to local, latest release to local, 3.2.0 to local are all passing)

In this run, only the first test (local to local) failed: https://github.com/atsign-foundation/sshnoports/actions/runs/5637947000 with the `Permission denied (publickey)` problem

But in this run, both the local to local test failed, AND the v3.2.0 to local test failed for the same (`Permission denied (publickey)`) reason https://github.com/atsign-foundation/sshnoports/actions/runs/5637996683/attempts/1
But then it passed in the re-run https://github.com/atsign-foundation/sshnoports/actions/runs/5637996683

The local-to-local test will not pass no matter how many times I re-run it

@JeremyTubongbanua Can you think of anything particularly special about the local-local test with regards to timings, setup, etc?

JeremyTubongbanua commented:
> Applied the fix mentioned in the previous comment, and now only one of the e2e tests is failing, but it's failing in a different way.
> 
> ```
> 130 sshnpd  | SHOUT|2023-07-23 18:19:51.546987| sshnpd |197de568-697a-4851-a789-9e82a722e5cb | Non-zero exit code from /usr/bin/ssh atsign@10.1.0.38 -p 40749 -t -t -i /tmp/.pem-436bdbbb-5052-4cb8-bd16-80ab0f86c376 -R 34925:localhost:22 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null sleep 15 
> 131 sshnpd  | SHOUT|2023-07-23 18:19:51.548265| sshnpd |197de568-697a-4851-a789-9e82a722e5cb | stdout   :  
> 132 sshnpd  | SHOUT|2023-07-23 18:19:51.548293| sshnpd |197de568-697a-4851-a789-9e82a722e5cb | stderr   : Warning: Permanently added '[10.1.0.38]:40749' (ED25519) to the list of known hosts.
> 133 sshnpd  | Load key "/tmp/.pem-436bdbbb-5052-4cb8-bd16-80ab0f86c376": error in libcrypto
> 134 sshnpd  | atsign@10.1.0.38: Permission denied (publickey).
> ```
> 
> I _think_ the repeat failure of the local-to-local test is an artefact of the test rig
> 
> All of the other tests which are running sshnpd from the branch (local) are passing (i.e. trunk to local, latest release to local, 3.2.0 to local are all passing)
> 
> In this run, only the first test (local to local) failed: https://github.com/atsign-foundation/sshnoports/actions/runs/5637947000 with the `Permission denied (publickey)` problem
> 
> But in this run, both the local to local test failed, AND the v3.2.0 to local test failed for the same (`Permission denied (publickey)`) reason https://github.com/atsign-foundation/sshnoports/actions/runs/5637996683/attempts/1 But then it passed in the re-run https://github.com/atsign-foundation/sshnoports/actions/runs/5637996683
> 
> The local-to-local test will not pass no matter how many times I re-run it
> 
> @JeremyTubongbanua Can you think of anything particularly special about the local-local test with regards to timings, setup, etc?

Yes, @XavierChanth recently refactored the tests and I believe the bug is addressed in a comment in this ticket - https://github.com/atsign-foundation/sshnoports/issues/276

JeremyTubongbanua commented:
Fix should be in #280

gkc commented:
@JeremyTubongbanua thanks - the problem I was encountering was actually a different problem (`Permission denied (publickey)`) which @cconstab diagnosed [here](https://github.com/atsign-foundation/sshnoports/issues/273#issuecomment-1647083625)

JeremyTubongbanua commented:
> @JeremyTubongbanua thanks - the problem I was encountering was actually a different problem (`Permission denied (publickey)`) which @cconstab diagnosed [here](https://github.com/atsign-foundation/sshnoports/issues/273#issuecomment-1647083625)

Awesome :)

gkc committed:
feat: added more options to the ssh command when we exec it
feat: add `-o StrictHostKeyChecking=accept-new` to the command that sshnp outputs

gkc committed:
Merge remote-tracking branch 'origin/trunk' into gkc-exec-ssh-on-host

gkc committed:
fix: used Process.start instead of Process.run

gkc committed:
fix: re-added option I'd removed

gkc committed:
refactor: Remove unused var from SSHNP

gkc committed:
refactor: Removed dead code from SSHNPDImpl

gkc committed:
refactor: Removed dead code from SSHNPDImpl

gkc committed:
feat: add '-o IdentitiesOnly=yes' to the ssh command which SSHNPImpl generates

gkc committed:
feat: add common/supported_ssh_clients.dart

gkc committed:
feat: SSHNPImpl now logs stdOut and stdErr of the ssh command when it execs ssh on the host

gkc committed:
feat: SSHNPDImpl: When exec'ing: `-o LogLevel=VERBOSE` instead of `-v`; `-o StrictHostKeyChecking` now set to `accept-new`, and removed `-o UserKnownHostsFile=/dev/null`

gkc committed:
feat: SSHNPDParams : add --ssh-client option and sshClient instance variable
refactor: made the SSHNPDParams args tests a bit easier to read
test: Add tests for --ssh-client option

gkc committed:
feat: added `final SupportedSshClient sshClient` to SSHNPD and SSHNPDImpl classes, and their factories / constructors
feat: implemented feature for new command-line flag --ssh-client so that it will use either the host's ssh command OR the pure dart SSHClient from dartssh2 package
build: re-added dartssh2 and ssh-key dependencies

gkc committed:
fix: removed unnecessary delay
refactor: made the creation of the args easier to read and not possible to break by accidentally omitting a comma
style: ran dart format

gkc commented:
@cconstab Ready for final review
